### DISCUSSION TITLE
Routing with Dynamic Params

### DISCUSSION THREAD

**Discussion by BearCooder** [vercel/next.js#55207] [App Router] [answered]
Hello.

I have a React Vite project and now I am in the process of migrating to Next.js and using the App Router. It is my first time I use Next js and its going good but I struggle with one thing.
I have the following route: `localhost:3000//gallery/:id`

So I managed well to get it work with app router when I access the `/gallery` route.

But on the gallery page I have some pictures. Every image has its own id. And when I click on an image I get redirected into the `/:id` page with more details about the image. 

How it works right now is that I use the dedicated id of the image and add it to the url in my route.

Currently I achieve it this way with wouter:

In the `/gallery` component:

```
import { Fragment, useState } from "react";
import { Link } from "wouter";

export default function Gallery() {
  const [index, setIndex] = useState<number | null>(null);

  const setActive = (id: number) => {
    setIndex(id);
  };

  return (
    <div className="flex flex-wrap justify-center">
      {artworks.map((artwork) => {
        index === artwork.id;
        return (
          <Fragment key={artwork.id}>
            <div className="px-5 py-5">
              <Link
                href={`gallery/${artwork.id}`}
                onClick={() => setActive(artwork.id)}
              >
                  <img
                    src={artwork.Image}
                    title={artwork.title}
                    alt={artwork.title}
                  />
              </Link>
            </div>
          </Fragment>
        );
      })}
    </div>
  );
}
```

And then in the component where I display the details and add the `/:id` to the url:

```
import { Link, useRoute } from "wouter";

function ArtworkDetails() {
  // Using wouter's useRoute to get the params from the route
  const [match, params] = useRoute("/gallery/:id");
  if (!match || !params.id) return <div>Artwork not found</div>;

  const id = parseInt(params.id, 10);
  const artwork = artworks.find((art) => art.id === id);

  if (!artwork) return <div className="text-white">Artwork not found</div>;

  return (
    <>
Some details..
    </>
  );
}
```

Here, we are using useRoute from wouter to retrieve the match and params from the current route. And match is a boolean indicating if the current route matches our pattern (/gallery/:id). And params is an object containing any dynamic segments of the route, in this case id. And with parseInt we are converting the id from the URL to an integer and then using that id to fetch the related artwork from the artworks array.

With wouter this works but I am struggling to implement this behaviour with the app router, I was looking in the documentation for similar examples but did not find any? 

Can someone help me please?

**Comment by icyJoseph** [comment]
You need to use `next.js` Link component. SPA routing is a bit fragile, if you don't use the Link given by the framework, then they won't be able to track the URL changes properly. You can't and should not mix routers.

You also need a `gallery/[id]/page.tsx` page, which can use either:

- Page params, https://nextjs.org/docs/app/api-reference/file-conventions/page, or
- useParams hook https://nextjs.org/docs/app/api-reference/functions/use-params

Note that the first is a Page Component, and the second is a Client Component

**Reply by BearCooder** [reply]
Thank you. Yes I want to replace wouter with the NextJs App Router to not mix routers.
I already have the `gallery/[id]/page.tsx` page and what I think I basically need for the same logic is like in this template example from Vercel: https://github.com/vercel/next.js/blob/canary/examples/with-cloudinary/pages/index.tsx

```
...
const router = useRouter()
  const { photoId } = router.query
  const [lastViewedPhoto, setLastViewedPhoto] = useLastViewedPhoto()

  const lastViewedPhotoRef = useRef<HTMLAnchorElement>(null)

  useEffect(() => {
    // This effect keeps track of the last viewed photo in the modal to keep the index page in sync when the user navigates back
    if (lastViewedPhoto && !photoId) {
      lastViewedPhotoRef.current.scrollIntoView({ block: 'center' })
      setLastViewedPhoto(null)
    }
  }, [photoId, lastViewedPhoto, setLastViewedPhoto])
...
```

But this example is with pages and there is no useRouter() in the App Router. So I cannot do
 ```
const router = useRouter()
const { photoId } = router.query
```

**Reply by GabenGar** [reply]
You have to use [`useParams()`](https://nextjs.org/docs/app/api-reference/functions/use-params) instead.

**Reply by BearCooder** [reply]
Hello and thank you very much!

I solved it this way:

In my Gallery page where I show all images I did the following:

```
export default function Gallery({
  searchParams,
}: {
  searchParams: { [key: string]: string | string[] | undefined };
}) {
  const selectedImage = searchParams.id as unknown | number;

  return (
    <div className="flex flex-wrap justify-center">
      {images.map((image) => {
        selectedImage === image.id;
        return (
          <Fragment key={image.id}>
            <div className="px-5 py-5">
              <Link href={`/gallery/${image.id}`}>
                  <Image
                    src={image.coverImage}
                    title={image.title}
                    alt={{image.title}
                    placeholder="blur"
                  />

```

The Link and Image are imported from Next. With searchParams I match it with the id that I have for every image. Its a number.
Then when I click on the image I get forwarded to the page gallery/[id].

As I am new to Next.JS I would be very happy to receive some feedback from you @GabenGar and @icyJoseph @leerob ?
It works, but is it fine as it is or are there better ways? Because I did not use [useParams()](https://nextjs.org/docs/app/api-reference/functions/use-params)?
I created a minimal working example sandbox with this logic, you can try and see it here: 
https://codesandbox.io/p/sandbox/falling-waterfall-8zn334?file=/app/page.tsx

Thank you again!

**Reply by BearCooder** [reply]
I know everyone is busy, but if you have a moment I would appreciate your input about my solution :)

**Reply by GabenGar** [reply]
What exactly do you want to do? Your current implementation will jerk the servers off constantly because you aren't rendering the page on build. Also you have to validate params properly instead of coercing the type.

**Reply by BearCooder** [reply]
This is what I want to do: I have the `/gallery` page, here I show from every photo a main image - the `coverImage`. Then when I click on a photo I want to display on the `/gallery/[id]` page more photos of the same kind. Thats what I tried to show with my working example on codesandbox.

**Your current implementation will jerk the servers off constantly because you aren't rendering the page on build** - Is this the case because my page in [id] is marked as "use client"? So would it be better to implement and import the clicking to next/previous image in another component and remove here the "use client?

**Also you have to validate params properly instead of coercing the type.** Sorry can you please elaborate?

**Reply by GabenGar** [reply]
> Is this the case because my page in [id] is marked as "use client"?

No, because you are using `searchParams` of the server component without [`generateStaticParams()`](https://nextjs.org/docs/app/api-reference/functions/generate-static-params), so it means the page is going to be rerendered on each request.
Either add `generateStaticParams()` or get them from `useParams()`.

>Sorry can you please elaborate?

Do you know what URLs are?

**Reply by BearCooder** [reply]
Ah ok I did not know this, thank you. I will look into [generateStaticParams()](https://nextjs.org/docs/app/api-reference/functions/generate-static-params) and either add or get it from `useParams()`.

And yes I know what URLs are, sorry if the question is dumb. I will try to figure it out.

**Reply by GabenGar** [reply]
> And yes I know what URLs are

Right, and what happens when your code accepts the URL like `localhost:3000//gallery/null`? What is the value of `searchParams.id` in this case?

**Reply by BearCooder** [reply]
My `const selectedImage = searchParams.id as unknown | number;` did not make sense. Since I want selectedImage as Number type, its better to do it runtime rather than forcing the type on TypeScript via this way: `const selectedImage = Number(searchParams.id)`. But in case searchParams.id is not available, I can use a dummy value so I wont get NaN, e.g.` const selectedImage = Number(searchParams.id ?? -1)`

So with this approach I think I would validate params properly?

**Reply by GabenGar** [reply, marked as answer]
If `searchParams.id` is not available then the entire page should redirect to a 404 page, since the logic of the page requires it to be present. The value validation logic depends on the context, in this specific example the values are literally in the specific range so you only need to check that the value is an integer and is no less than 0 (btw don't start serial IDs with zeroes as it brings a lot of problems down the line) and no more than maximum ID. It still doesn't guarantee that the item with this ID will be found, but it filters out obviously wrong inputs.
